### PR TITLE
add ui template for ctl

### PR DESCRIPTION
**Pull request by Alexanderlacuna**
#### Description

This PR integrate to gn3 ctl and making the required plots

### CODE REVIEW

File: wqflask/wqflask/ctl/gn3_ctl_analysis.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+from utility.tools import locate
+
+from base.trait import create_trait
+from base.trait import retrieve_sample_data
+from base import data_set
+
+
+def process_significance_data(significant_data):
+    """function to process significance the data for
+    datatables"""
```
Comment (BonfaceKilz) [inline wqflask/wqflask/ctl/gn3_ctl_analysis.py]: *issue: nitpick:* Pick better names; and try to have your docstrings more clearer.  `significance_data` from what I can see, would perhaps simply be "dataset".  Names of functions and objects should reflect precisely what they represent.  ATM, your docstrings, and to some extent, function names are pretty opaque.

File: wqflask/wqflask/ctl/gn3_ctl_analysis.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                dataset_rows[index].append(col_data)
+
+    return {
+        "col_names": col_names,
+        "data_set_rows": dataset_rows
+    }
+
+
+def parse_geno_data(dataset_group_name) -> dict:
+    """function to parse geno file data"""
```
Comment (BonfaceKilz) [inline wqflask/wqflask/ctl/gn3_ctl_analysis.py]: **issue:** We have a function "parse_geno_data"; and it's docstring: "function to parse geno file data".  They say the same thing.  If the docstring isn't useful, don't put it there.

File: wqflask/wqflask/ctl/gn3_ctl_analysis.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        "genotypes": list(itertools.chain(*markers)),
+        "markernames": markernames,
+        "individuals": parser.individuals
+
+
+    }
+
+
+def parse_phenotype_data(trait_db_list, dataset, individuals):
+    """function to parse and generate phenodata"""
```
Comment (BonfaceKilz) [inline wqflask/wqflask/ctl/gn3_ctl_analysis.py]: **issue:** Same issue above with naming.

File: wqflask/wqflask/views.py [hunk 4]
```diff
@@ -354,6 +357,22 @@ def ctl_setup():
     return render_template("ctl_setup.html", **request.form)
 
 
+
+@app.route("/ctl_results",methods=["POST"])
+def ctl_results():
+
+    ctl_results = run_ctl(request.form)
+    return render_template("gn3_ctl_results.html",**ctl_results)
```
Comment (BonfaceKilz) [inline wqflask/wqflask/views.py]: *nitpick*:  
```suggestion
    return render_template("gn3_ctl_results.html", **ctl_results)
```

File: wqflask/wqflask/views.py [hunk 5]
```diff
@@ -354,6 +357,22 @@ def ctl_setup():
...
+
+
+@app.route("/ctl_network_files/<file_name>/<file_type>")
+def fetch_network_files(file_name,file_type):
```
Comment (BonfaceKilz) [inline wqflask/wqflask/views.py]: *nitpick:*
```suggestion
def fetch_network_files(file_name, file_type):
```

**Comment by Alexanderlacuna** [comment]
@BonfaceKilz  thanks for the review, need to work on the naming of my variables in general.Will ping you when I fix the above.

**Comment by Alexanderlacuna** [comment]
plus before merge, will set things up for  @robwwilliams @zsloan   to test this on penguin and give their feedback

**Comment by BonfaceKilz** [comment]
Alexander Kabui ***@***.***>
anaandika:
> @BonfaceKilz  thanks for the review, need to work on the naming of my variables in general.Will ping you when I fix the above.
>

Sure!

**Comment by BonfaceKilz** [comment]
+1

File: wqflask/wqflask/ctl/gn3_ctl_analysis.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    return {
+        "col_names": col_names,
+        "data_set_rows": dataset_rows
+    }
+
+
+def parse_geno_data(dataset_group_name) -> dict:
+    """
+    Args:
+        dataset_group_name: string name
```
Comment (BonfaceKilz) [inline wqflask/wqflask/ctl/gn3_ctl_analysis.py:32]: **nitpick:**  Args should be more descriptive.  And types captures what you've documented as an arg:

```
def parse_geno_data(dataset_group_name -> str) -> dict:
...
```

I think I'll have to look for resources to point you to show you how to improve documentation.  @arunisaac if you have anything in mind, you could share ;)

File: wqflask/wqflask/ctl/gn3_ctl_analysis.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    form_data["nperm"] = int(form_data["nperm"])
+    form_data["significance"] = float(form_data["significance"])
+    form_data["strategy"] = form_data["strategy"].capitalize()
+
+    return form_data
+
+
+def run_ctl(requestform):
+    """function to make an api call
+    to gn3 and run ctl"""
```
Comment (BonfaceKilz) [inline wqflask/wqflask/ctl/gn3_ctl_analysis.py:102]: **nitpick:** We already know this is a function.

```suggestion
    """Make an api call to gn3 and run ctl"""
```

**Review by BonfaceKilz — APPROVED** [review]
LGTM.  @Alexanderlacuna should I go ahead and merge this?  You can address minor issues in another PR.

**Comment by arunisaac** [comment]
> I think I'll have to look for resources to point you to show you how
> to improve documentation.  @arunisaac if you have anything in mind,
> you could share ;)

I haven't really thought much about docstrings, and don't have any
resources in mind at the moment. I'm hoping you will come up with
something, Bonface! :-)

**Comment by BonfaceKilz** [comment]
Arun Isaac ***@***.***> anaandika:

>> I think I'll have to look for resources to point you to show you how
>> to improve documentation.  @arunisaac if you have anything in mind,
>> you could share ;)
>
> I haven't really thought much about docstrings, and don't have any
> resources in mind at the moment. I'm hoping you will come up with
> something, Bonface! :-)

Cool.  That's one of the things I want improved
with how we code.  Standard ways of documenting
things.

-- 
Bonface M. K.
D4F09EB110177E03C28E2FE1F5BBAE1E0392253F (hkp://keys.gnupg.net)
Free Software Activist
Humble GNU Emacs User | Bearer of scheme-y parens
Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz

**Comment by Alexanderlacuna** [comment]
@BonfaceKilz  This is  can be merged

**Pull request merged**